### ISSUE TITLE
about the time

### ISSUE THREAD

**Issue by OverDriveMC** [horovod/horovod#372]
I want to see the performance of horovod, so I try to run the examples/tensorflow_mnist.py,

First,I run the command:    

    time mpirun -np 1 \  
    -H localhost:1 \
    -bind-to none -map-by slot \
    -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH \
    -mca pml ob1 -mca btl ^openib \
    python  tensorflow_mnist.py
then the log like this:

>INFO:tensorflow:loss = 0.0296682, step = 1911 (0.077 sec)
>INFO:tensorflow:loss = 0.00212721, step = 1921 (0.072 sec)
>INFO:tensorflow:loss = 0.0195793, step = 1931 (0.072 sec)
>INFO:tensorflow:loss = 0.0365237, step = 1941 (0.071 sec)
>INFO:tensorflow:loss = 0.0026851, step = 1951 (0.074 sec)
>INFO:tensorflow:loss = 0.01054, step = 1961 (0.074 sec)

>real	0m25.062s
>user	0m25.384s
>sys	0m7.820s

Then I try two gpu:  

    time mpirun -np 2 \  
    -H localhost:2 \
    -bind-to none -map-by slot \
    -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH \
    -mca pml ob1 -mca btl ^openib \
    python  tensorflow_mnist.py
the log like this:

> INFO:tensorflow:loss = 0.0024739, step = 981 (0.385 sec)
>INFO:tensorflow:loss = 0.0357948, step = 981 (0.384 sec)
>INFO:tensorflow:loss = 0.0278996, step = 991 (0.384 sec)
>INFO:tensorflow:loss = 0.0805988, step = 991 (0.387 sec)

>real	0m48.151s
>user	1m22.616s
>sys	0m32.004s

I wonder why the performance one gpu better than two gpu

**Comment by tgaddair** [comment]
Hey @OverDriveMC, I think what you're observing is the overhead of using interprocess communication with a relatively small training dataset.

Here are the results I had running on 1 GPU:

```
INFO:tensorflow:step = 19962, loss = 0.0 (0.039 sec)
INFO:tensorflow:step = 19972, loss = 0.0 (0.041 sec)
INFO:tensorflow:step = 19982, loss = 0.0 (0.041 sec)
INFO:tensorflow:step = 19992, loss = 0.0 (0.040 sec)
INFO:tensorflow:Saving checkpoints for 20000 into ./checkpoints/model.ckpt.

real	0m56.721s
user	0m58.512s
sys	0m13.724s
```

And like you saw, the time between every 10 steps was significantly higher with 2 GPUs:

```
INFO:tensorflow:step = 9980, loss = 0.0001958076 (0.275 sec)
INFO:tensorflow:step = 9980, loss = 0.0 (0.275 sec)
INFO:tensorflow:step = 9990, loss = 0.041696638 (0.258 sec)
INFO:tensorflow:step = 9990, loss = 0.0 (0.260 sec)
INFO:tensorflow:Saving checkpoints for 10000 into ./checkpoints/model.ckpt.

real	4m30.630s
user	8m58.624s
sys	0m43.456s
```

However, when I scale up the number of nodes by 8x (16 processes total), you see that the additional overhead only goes up by 2x:

```
Thu Jul 12 15:20:46 2018[1,1]<stderr>:INFO:tensorflow:step = 1240, loss = 1.3232173e-07 (0.464 sec)
Thu Jul 12 15:20:46 2018[1,12]<stderr>:INFO:tensorflow:step = 1240, loss = 0.00065653084 (0.462 sec)
Thu Jul 12 15:20:46 2018[1,3]<stderr>:INFO:tensorflow:step = 1240, loss = 6.882769e-05 (0.466 sec)
Thu Jul 12 15:20:46 2018[1,2]<stderr>:INFO:tensorflow:step = 1240, loss = 9.430028e-05 (0.466 sec)
Thu Jul 12 15:20:46 2018[1,0]<stderr>:INFO:tensorflow:Saving checkpoints for 1250 into ./checkpoints/model.ckpt.

real	1m10.399s
user	0m0.128s
sys	0m0.180s
```

But I think the crux of it is that the dataset is just too small in this example to give you a meaningful benchmark.  Using Horovod in this instance would mostly be for illustrative purposes.

**Comment by OverDriveMC** [comment]
OK，Thanks~
another question, I wonder  whether nv_peer_memory will play an important role in the speed?
because I may not have permission to install it on the server.

**Comment by Stonesjtu** [comment]
@OverDriveMC I think peer memory access doesn't require any software dependencies, and it's (probably) opened by default.
You can run with CUDA sample namely `p2pbandwidthtest`
